packages/web3-providers-ws/src/index.ts

```diff
@@ -0,0 +1,387 @@
+import { EventEmitter } from 'events';
+import {
+	Web3BaseProvider,
+	Web3BaseProviderStatus,
+	JsonRpcId,
+	JsonRpcPayload,
+	JsonRpcResponse,
+	JsonRpcResult,
+	Web3BaseProviderCallback,
+	ConnectionEvent,
+	SubscriptionResultNotification,
+} from 'web3-common';
+import { IMessageEvent, w3cwebsocket as W3WS } from 'websocket';
+import {
+	InvalidConnectionError,
+	ConnectionTimeoutError,
+	PendingRequestsOnReconnectingError,
+	ConnectionNotOpenError,
+	InvalidClientError,
+} from 'web3-common/dist/errors';
+import { WebSocketOptions, WSRequestItem } from './types';
+import { DeferredPromise } from './deferredPromise';
+
+export default class WebSocketProvider extends Web3BaseProvider {
+	private readonly wsEventEmitter: EventEmitter = new EventEmitter();
+
+	private readonly clientUrl: string;
+	private readonly wsProviderOptions: WebSocketOptions;
+
+	private webSocketConnection?: W3WS;
+
+	/* eslint-disable @typescript-eslint/no-explicit-any */
+	private readonly requestQueue: Map<JsonRpcId, WSRequestItem<any, any>>;
+	/* eslint-disable @typescript-eslint/no-explicit-any */
+	private readonly sentQueue: Map<JsonRpcId, WSRequestItem<any, any>>;
+
+	private providerStatus: Web3BaseProviderStatus;
+
+	private reconnectAttempts!: number;
+	private lastDataChunk!: string;
+	private lastChunkTimeout!: NodeJS.Timeout;
+
+	public constructor(clientUrl: string, wsProviderOptions?: WebSocketOptions) {
+		super();
+		if (!WebSocketProvider.validateProviderUrl(clientUrl))
+			throw new InvalidClientError(clientUrl);
+
+		this.clientUrl = clientUrl;
+		if (wsProviderOptions !== undefined) this.wsProviderOptions = wsProviderOptions;
+		else
+			this.wsProviderOptions = {
+				customTimeout: 1000 * 15,
+			};
+
+		if (!this.wsProviderOptions.customTimeout) this.wsProviderOptions.customTimeout = 1000 * 15;
+
+		if (!this.wsProviderOptions.reconnectOptions)
+			this.wsProviderOptions.reconnectOptions = {
+				auto: false,
+				delay: 5000,
+				maxAttempts: 5,
+				onTimeout: false,
+			};
```

Comment:
The way this is currently written doesn't validate that all these properties are going to get set. Instead I'd do something like:

```typescript
const DEFAULT_WS_PROVIDER_OPTIONS = {
    customTimeout: 1000 * 15,
    reconnectOptions: {
        auto: false,
        delay: 5000,
        maxAttempts: 5,
        onTimeout: false,
    }
}

this.wsProviderOptions = {
    ...DEFAULT_WS_PROVIDER_OPTIONS,
    ...wsProviderOptions
    reconnectOptions: {
        ...DEFAULT_WS_PROVIDER_OPTIONS.reconnectOptions,
        ...wsProviderOptions?.reconnectOptions
    }
}
```

This will default all the needed properties regardless of how the user passes `wsProviderOptions`